### PR TITLE
Added Custom Events, Time Profiling and Set Data How To's from FAQs

### PR DESCRIPTION
**Pull request by Priyansi**

### CODE REVIEW

**Comment by sdesrozis** [comment]
@Priyansi Thank you. I think there is a lack of explanations in the notebooks, I mean some context, and examples sound very basic, maybe too useless or without context. Could we improve the content by giving some more concrete illustration ?

**Comment by Priyansi** [comment]
> @Priyansi Thank you. I think there is a lack of explanations in the notebooks, I mean some context, and examples sound very basic, maybe too useless or without context. Could we improve the content by giving some more concrete illustration ?

@sdesrozis I agree. What are your thoughts on dividing How-Tos in 2 parts -
* A TLDR at the top of the notebook that provides just the code without any context or dummy data
* We can set up a basic model, use MNIST data and other parameters to give these how-to's more context.

**Comment by sdesrozis** [comment]
Having just the code could be a bit hard to understand for a beginner. I would have liked something educational like what is done in the PyTorch examples and tutorials. Even if we are discussing a few basic features, maybe the code without explanation could not be sufficient.

Btw, having a concrete example at the end or not is a good idea. 

As an inspiration, not on the content but in the explanatory approach, here is an interesting link

https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html

What do you think ?

**Comment by Priyansi** [comment]
Yes, I understand. However, that would make the how-to guides section into tutorials. However, the purpose of How-To Guides is to solve a specific problem with a minimal explanation as mentioned [here](https://diataxis.fr/how-to-guides/). How-To Guides aren't meant for complete beginners. We will have tutorials covering custom events and time profiling along with other Ignite concepts but if a user wanted a quick way to get the code for these concepts, they will turn to how-to guides. The autograd example you mentioned is definitely a tutorial (with a mix of explanations) and putting them in how-to guides would defeat this section's purpose. 

If we want to an in-depth explanation of certain concepts, we can add them in pytorch-ignite/pytorch-ignite.github.io/explanations

What do you think?

**Comment by sdesrozis** [comment]
Ok, I didn't get that howto involves minimal explanation ! Now it's quite clear, thanks for the clarification ! Therefore I agree, no need to consider my remarks, sorry for my misunderstanding.

File: how-to-guides/custom-events-pass.ipynb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+   "source": [
+    "# How to create Custom Events based on Forward or Backward Pass"
+   ],
+   "metadata": {}
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "Ignite provides you the flexibility to add events based on the loss\n",
+    "calculation and backward pass as follows:"
```
Comment (vfdev-5) [inline how-to-guides/custom-events-pass.ipynb:14]: This sounds like custom events helps with loss calculation and backward pass. Please rephrase this sentence as custom events is a generic concept and here it is an example

File: how-to-guides/custom-events-pass.ipynb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    "    \"\"\"\n",
+    "    BACKWARD_STARTED = 'backward_started'\n",
+    "    BACKWARD_COMPLETED = 'backward_completed'\n",
+    "    OPTIM_STEP_COMPLETED = 'optim_step_completed'\n",
+    "\n",
+    "def update(engine, batch):\n",
+    "    model.train()\n",
+    "    opitmizer.zero_grad()\n",
+    "    x, y = process_batch(batch)\n",
+    "    y_pred = model(x)\n",
```
Comment (vfdev-5) [inline how-to-guides/custom-events-pass.ipynb:36]: Let's make this code snippet self-contained and executable

File: how-to-guides/switch-data-training.ipynb [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    "# How to switch data provider during training"
+   ],
+   "metadata": {}
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "User can easily switch data provider during the training using\n",
+    "[`set_data()`](https://pytorch.org/ignite/v0.4.5/generated/ignite.engine.engine.Engine.html#ignite.engine.engine.Engine.set_data). See an example in the\n",
+    "documentation of the method."
```
Comment (vfdev-5) [inline how-to-guides/switch-data-training.ipynb:15]: > See an example in the documentation of the method.

Why we need to send user somewhere else. Let's make a self-contained how-to guide

File: how-to-guides/switch-data-training.ipynb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "source": [
+    "from ignite.engine import Engine, Events\n",
+    "\n",
+    "data1 = [1, 2, 3]\n",
+    "data2 = [4, 5, 6]\n",
```
Comment (vfdev-5) [inline how-to-guides/switch-data-training.ipynb:26]: To make the logging more clear that we changed the data previously I used 
```
data2 = [10 + a, 10 +b, 10 + c]
```
such we clearly see the differences between batches

File: how-to-guides/custom-events-pass.ipynb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    "    print(\"Hello\")"
+   ],
+   "outputs": [],
+   "metadata": {}
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "For a detailed implementation, read [TBPTT\n",
+    "Trainer](https://pytorch.org/ignite/master/_modules/ignite/contrib/engines/tbptt.html#create_supervised_tbptt_trainer)."
```
Comment (vfdev-5) [inline how-to-guides/custom-events-pass.ipynb:64]: Please make the links without explicitly using master or v0.4.5 everywhere.

File: how-to-guides/time-profiling.ipynb [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+   ],
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "\n",
+      "----------------------------  -------------------  --------------  --------------  --------------  --------------  --------------  \n",
+      "Handler                       Event Name                 Total(s)      Min(s)/IDX      Max(s)/IDX         Mean(s)          Std(s)  \n",
+      "----------------------------  -------------------  --------------  --------------  --------------  --------------  --------------  \n",
```
Comment (vfdev-5) [inline how-to-guides/time-profiling.ipynb:241]: The output does not look nice. Let's find a way to improve it

File: how-to-guides/time-profiling.ipynb [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+    "# For every datapoint, trainer will sleep for 1 second\n",
+    "trainer = Engine(lambda engine, batch: time.sleep(1))\n",
+    "\n",
+    "# Create an object of the profiler and attach an engine to it\n",
+    "profiler = HandlersTimeProfiler()\n",
+    "profiler.attach(trainer)   \n",
+    "\n",
+    "@trainer.on(Events.EPOCH_COMPLETED(every=1))\n",
+    "def log_intermediate_results():\n",
+    "    profiler.print_results(profiler.get_results())\n",
```
Comment (vfdev-5) [inline how-to-guides/time-profiling.ipynb:229]: Let's add more handlers to show how it works

File: how-to-guides/time-profiling.ipynb [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+   ],
+   "metadata": {}
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "## Custom time measures\n",
+    "\n",
+    "Custom time measures can be performed using [`Timer`](https://pytorch.org/ignite/v0.4.5/generated/ignite.handlers.timing.Timer.html#timer). See its\n",
+    "docstring for details."
```
Comment (vfdev-5) [inline how-to-guides/time-profiling.ipynb:309]: Let's add the example of usage instead of sending to the docs

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @Priyansi 
I left some comments to update the content before merging.

**Comment by Priyansi** [comment]
Hey @vfdev-5 I will remove the time-profiling how-to for now since we are breaking them into separate notebooks like in #6

**Comment by vfdev-5** [comment]
> Hey @vfdev-5 I will remove the time-profiling how-to for now since we are breaking them into separate notebooks like in #6 . Will that be alright?

@Priyansi the problem with the notebook from #6 is that it shows only 1 feature instead of 3 from this how-to guide

**Comment by Priyansi** [comment]
> @Priyansi the problem with the notebook from #6 is that it shows only 1 feature instead of 3 from this how-to guide

Yes, I will open two new issues for the other types.

File: how-to-guides/custom-events-pass.ipynb (hunk 2 above)
Comment (trsvchn) [reply how-to-guides/custom-events-pass.ipynb:36]: Agree! That's why we use notebooks for this

**Comment by vfdev-5** [comment]
@Priyansi what remains here to finalize this PR ?

**Comment by Priyansi** [comment]
@vfdev-5 I have updated both notebooks and the PR should be good to merge. Again, I apologize for the delay.

cc @trsvchn @sdesrozis
